File: src/dw_conv.cpp
#include <iostream>
#include <fstream>
#include <vector>
#include <cmath>

// **Depthwise Convolution Configuration**
#define IN_C 3       // Input channels (same as output channels)
#define H_IN 10      // Input height (including padding)
#define W_IN 12      // Input width (including padding)
#define K 3          // Kernel size
#define STRIDE 2     // Stride
#define H_OUT ((H_IN - K) / STRIDE + 1)  // Output height
#define W_OUT ((W_IN - K) / STRIDE + 1)  // Output width

typedef float fm_t;

// **Depthwise Convolution Function**
void dw_conv3x3(
    fm_t outBuffer[IN_C][H_OUT][W_OUT],
    fm_t inBuffer[IN_C][H_IN][W_IN], 
    fm_t weights[IN_C][K][K], 
    fm_t bias[IN_C]
) {
    for (int c = 0; c < IN_C; c++) {  
        for (int i = 0; i < H_OUT; i++) {  
            for (int j = 0; j < W_OUT; j++) {  
                fm_t sum = bias[c];  // Start with bias
                for (int ki = 0; ki < K; ki++) {  
                    for (int kj = 0; kj < K; kj++) {  
                        int in_i = i * STRIDE + ki;  
                        int in_j = j * STRIDE + kj;  
                        sum += inBuffer[c][in_i][in_j] * weights[c][ki][kj];  
                    }
                }
                outBuffer[c][i][j] = sum;  // Store result in output buffer
            }
        }
    }
}

// **Load data from file**
void load_data(const std::string &filename, std::vector<fm_t> &data) {
    std::ifstream file(filename);
    if (!file) {
        std::cerr << "Cannot open file: " << filename << std::endl;
        exit(1);
    }
    fm_t value;
    while (file >> value) {
        data.push_back(value);
    }
}

int main() {
    // **Create Buffers**
    fm_t inBuffer[IN_C][H_IN][W_IN];
    fm_t weights[IN_C][K][K];
    fm_t bias[IN_C];
    fm_t outBuffer[IN_C][H_OUT][W_OUT];

    // **Load Data**
    std::vector<fm_t> in_data, weight_data, bias_data;
    load_data("data/inBuffer_dwconv.txt", in_data);
    load_data("data/weights_dwconv.txt", weight_data);
    load_data("data/bias_dwconv.txt", bias_data);

    // **Fill inBuffer**
    int index = 0;
    for (int c = 0; c < IN_C; c++)
        for (int i = 0; i < H_IN; i++)
            for (int j = 0; j < W_IN; j++)
                inBuffer[c][i][j] = in_data[index++];

    // **Fill weights**
    index = 0;
    for (int c = 0; c < IN_C; c++)
        for (int ki = 0; ki < K; ki++)
            for (int kj = 0; kj < K; kj++)
                weights[c][ki][kj] = weight_data[index++];

    // **Fill bias**
    for (int c = 0; c < IN_C; c++)
        bias[c] = bias_data[c];

    // **Perform Depthwise Convolution**
    dw_conv3x3(outBuffer, inBuffer, weights, bias);

    // **Write output to file**
    std::ofstream file("data/out_cpp_dwconv.txt");
    if (!file) {
        std::cerr << "Cannot write to file: data/out_cpp_dwconv.txt" << std::endl;
        return 1;
    }
    for (int c = 0; c < IN_C; c++)
        for (int i = 0; i < H_OUT; i++)
            for (int j = 0; j < W_OUT; j++)
                file << outBuffer[c][i][j] << "\n";

    std::cout << "Depthwise Conv 3x3 computation completed, results saved to `data/out_cpp_dwconv.txt`" << std::endl;
    return 0;
}

File: src/dw_conv.py
import torch
import numpy as np
import os

# **Parameter Configuration**
IN_DEPTH = 3
H_IN, W_IN = 10, 12  # Input size including padding

# **Ensure 'data' directory exists**
os.makedirs("data", exist_ok=True)

# **1. Generate random input data**
torch.manual_seed(42)
inBuffer = torch.randn(IN_DEPTH, H_IN, W_IN, dtype=torch.float32)  # Input feature map
weights = torch.randn(IN_DEPTH, 3, 3, dtype=torch.float32)  # 3x3 Depthwise kernel
bias = torch.randn(IN_DEPTH, dtype=torch.float32)  # Bias

# **2. Compute Depthwise Convolution (stride=1, padding=0)**
conv = torch.nn.Conv2d(IN_DEPTH, IN_DEPTH, kernel_size=3, stride=2, padding=0, groups=IN_DEPTH, bias=True)
conv.weight.data = weights.unsqueeze(1)  # Depthwise conv weight format
conv.bias.data = bias
out_torch = conv(inBuffer.unsqueeze(0)).squeeze(0).detach().numpy()  # Remove batch dimension

# **3. Save data (.npy format)**
np.save("data/inBuffer_dwconv.npy", inBuffer.numpy())  # Input values
np.save("data/weights_dwconv.npy", weights.numpy())    # 3x3 Kernel
np.save("data/bias_dwconv.npy", bias.numpy())          # Bias
np.save("data/out_torch_dwconv.npy", out_torch)        # PyTorch results

# **4. Save data (.txt format)**
np.savetxt("data/inBuffer_dwconv.txt", inBuffer.numpy().flatten(), fmt="%.6f")
np.savetxt("data/weights_dwconv.txt", weights.numpy().flatten(), fmt="%.6f")
np.savetxt("data/bias_dwconv.txt", bias.numpy(), fmt="%.6f")
np.savetxt("data/out_torch_dwconv.txt", out_torch.flatten(), fmt="%.6f")

print("Depthwise Conv 3x3 input & output data saved to `data/` directory.")
